runtime/connectors/s3/s3.go

```diff
@@ -72,46 +77,49 @@ func (c connector) Spec() connectors.Spec {
 	return spec
 }
 
-func (c connector) ConsumeAsFile(ctx context.Context, env *connectors.Env, source *connectors.Source) (string, error) {
+func (c connector) ConsumeAsFile(ctx context.Context, env *connectors.Env, source *connectors.Source) ([]string, error) {
 	conf, err := ParseConfig(source.Properties)
 	if err != nil {
-		return "", fmt.Errorf("failed to parse config: %w", err)
+		return nil, fmt.Errorf("failed to parse config: %w", err)
 	}
 
-	// The session the S3 Downloader will use
-	sess, err := getAwsSessionConfig(conf)
-	if err != nil {
-		return "", fmt.Errorf("failed to start session: %w", err)
+	if !doublestar.ValidatePattern(conf.Path) {
+		// ideally this should be validated at much earlier stage
+		// keeping it here to have s3 specific validations
+		return nil, fmt.Errorf("glob pattern %s is invalid", conf.Path)
 	}
 
-	// Create a downloader with the session and default options
-	downloader := s3manager.NewDownloader(sess)
-
-	bucket, key, extension, err := awsURLParts(conf.Path)
+	bucket, glob, _, err := s3URLParts(conf.Path)
 	if err != nil {
-		return "", fmt.Errorf("failed to parse path %s, %w", conf.Path, err)
+		return nil, fmt.Errorf("failed to parse path %s, %w", conf.Path, err)
 	}
 
-	f, err := os.CreateTemp(
-		os.TempDir(),
-		fmt.Sprintf("%s*%s", source.Name, extension),
-	)
+	sess, err := getAwsSessionConfig(conf)
 	if err != nil {
-		return "", fmt.Errorf("os.Create: %w", err)
+		return nil, fmt.Errorf("failed to start session: %w", err)
 	}
-	defer f.Close()
 
-	// Write the contents of S3 Object to the f
-	_, err = downloader.DownloadWithContext(ctx, f, &s3.GetObjectInput{
-		Bucket: aws.String(bucket),
-		Key:    aws.String(key),
-	})
+	bucketObj, err := s3blob.OpenBucket(ctx, sess, bucket, nil)
 	if err != nil {
-		os.Remove(f.Name())
-		return "", fmt.Errorf("failed to download f, %w", err)
+		return nil, fmt.Errorf("failed to open bucket %s, %w", bucket, err)
 	}
+	defer bucketObj.Close()
 
-	return f.Name(), nil
+	fetchConfigs := rillblob.FetchConfigs{
+		MaxTotalSize:      conf.MaxTotalSize,
+		MaxMatchedObjects: conf.MaxMatchedObjects,
+		MaxObjectsListed:  conf.MaxObjectsListed,
+		PageSize:          conf.PageSize,
+	}
+	return rillblob.FetchFileNames(ctx, bucketObj, fetchConfigs, glob, bucket)
+}
+
+func s3URLParts(path string) (string, string, string, error) {
+	u, err := url.Parse(path)
+	if err != nil {
+		return "", "", "", err
+	}
+	return u.Host, strings.Replace(u.Path, "/", "", 1), fileutil.FullExt(u.Path), nil
```

Comment:
Is removing the initial `/` necessary? The old `awsURLParts` didn't do it. Wondering if it could have unintended side effects.

Reply:
Yes. We want the globpattern to be not starting with '/'